docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,541 @@
+  apply this to the parameters in a final step and return the updated parameters
+  and optimizer state.
+- We can use a `get_learning_rate()` that returns the learning rate depending on
+  the step number when defining the Optax gradient update transformation. Above
+  code illustrates how this can be a drop-in replacement for [previous Code],
+  where this update function already exists (notice how we need to invert the
+  sign because we add the gradient update to the parameters). See also [Optax
+  #20](https://github.com/deepmind/optax/issues/20) for an ongoing discussion
+  about scheduling hyper parameters.
+
+## Multi Optimizer
+[Multi Optimizer]: #multi-optimizer
+
+The [previous API] defined `flax.optim.MultiOptimizer` for processing different
+parts of the parameter tree with different optimizers:
+
+```python
+biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: path.endswith('/bias'))
+not_biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: not path.endswith('/bias'))
+
+optimizer_def = flax.optim.MultiOptimizer(
+    (biases_traversal, flax.optim.GradientDescent(learning_rate=0.1)),
+    (not_biases_traversal, flax.optim.GradientDescent(learning_rate=0.05)),
+)
+```
+
+Note how we first define a traversal that selects parameters based on their
+path (which is the concatenation of module scopes and variable name), and then
+create a `MultiOptimizer` that binds a different optimizer for each of these
+separate traversals.
+
+A similar scheme could be implemented with optimizers following the functional
+API:
+
+```python
+# flax/opt/base.py
+
+FilterFn = Callable[[str], bool]
+
+
+def multi_optimizer(
+  *filters_and_optimizers: Tuple[FilterFn, Optimizer]
+) -> Optimizer:
+  """Each `Optimizer` is restricted to params where `FilterFn` returns True."""
+
+  traversals_and_optimizers = [
+      (flax.optim.ModelParamTraversal(lambda path, _: filter_fn(path)), opt)
+      for filter_fn, opt in filters_and_optimizers
+  ]
+
+  def init_fn(params: Params) -> Sequence[OptimizerState]:
+    return [
+        opt.init(list(traversal.iterate(params)))
+        for traversal, opt in traversals_and_optimizers
+    ]
+
+  def update_fn(
+      grads: Gradients, params: Params, states: Sequence[OptimizerState]
+  ) -> Tuple[Params, Sequence[OptimizerState]]:
+    new_params = params
+    new_states = []
+    for (traversal, opt), state in zip(traversals_and_optimizers, states):
+      p = list(traversal.iterate(params))
+      g = list(traversal.iterate(grads))
+      new_p, new_state = opt.update(g, p, state)
+      new_params = traversal.set(new_p, new_params)
+      new_states.append(new_state)
+    return new_params, new_states
+
+  return Optimizer(init_fn, update_fn)
+```
+
+Remarks:
+
+- The `flax.optim.ModelParamTraversal` is reused.
+- The interface is a bit simplified, allowing to filter by path only (filtering
+  by value was never used), and shortening user code a bit by instantiating the
```

Comment:
filtering by shape and/or dtype might still be useful

Reply:
Re-added.